src/utilities/replaceASTVariables.js

```diff
@@ -0,0 +1,36 @@
+import type { ObjMap } from '../jsutils/ObjMap';
+
+import type { ValueNode, ConstValueNode } from '../language/ast';
+import { Kind } from '../language/kinds';
+import { visit } from '../language/visitor';
+
+import { valueToLiteral } from './valueToLiteral';
+
+/**
+ * Replaces any Variables found within an AST Value literal with literals
+ * supplied from a map of variable values, returning a constant value.
+ *
+ * Used primarily to ensure only complete constant values are used during input
+ * coercion of custom scalars which accept complex literals.
+ */
+export function replaceASTVariables(
+  valueNode: ValueNode,
+  variables: ?ObjMap<mixed>,
+): ConstValueNode {
+  return visit(valueNode, {
+    Variable(node) {
+      return valueToLiteral(variables?.[node.name.value]);
```

Comment:
we need to consider non provided variables values: a variables reference ast reference would need to completely removed I think to emulate a non existing value (and not just null)

Reply:
Correct, the section on "ObjectValue" below performs that task, the last case in replaceASTVariables-test.js tests the behavior

Reply:
👍